src/librustc_data_structures/array_vec.rs

```diff
@@ -36,13 +41,56 @@ pub struct ArrayVec<A: Array> {
     values: A::PartialStorage
 }
 
+impl<A> Clone for ArrayVec<A>
+    where A: Array,
+          A::Element: Clone {
+    fn clone(&self) -> Self {
+        let mut v = ArrayVec::new();
+        v.extend(self.iter().cloned());
+        v
+    }
+}
+
 impl<A: Array> ArrayVec<A> {
     pub fn new() -> Self {
         ArrayVec {
             count: 0,
             values: Default::default(),
         }
     }
+
+    pub fn len(&self) -> usize {
+        self.count
+    }
+
+    /// Panics when the stack vector is full.
+    pub fn push(&mut self, el: A::Element) {
+        unsafe {
+            let arr = &mut self.values as &mut [ManuallyDrop<_>];
+            arr[self.count].value = el;
+        }
+        self.count += 1;
+    }
+
+    pub fn pop(&mut self) -> Option<A::Element> {
+        if self.count > 0 {
+            let arr = &mut self.values as &mut [ManuallyDrop<_>];
+            let value = mem::replace(&mut arr[self.count - 1], ManuallyDrop::new());
```

Comment:
You don't need `mem::replace`, you can just do `ptr::read(&arr[self.count - 1].value)`.
Also, you can decrement first, and use `self.count` then.